File: ios_src/Shuiguang2/Shuiguang/HistoryModelOperation.m
//
//  HistoryModelOperation.m
//  Shuiguang
//
//  Created by dehualai on 3/22/17.
//  Copyright © 2017 NanoMed. All rights reserved.
//

#import "HistoryModelOperation.h"
#import "DBDetection+Operation.h"
#import "Util.h"

#define TIME_RANGE_LAST_WEEK 1
#define TIME_RANGE_LAST_TWO_WEEKS 2
#define TIME_RANGE_LAST_MONTH 3
#define TIME_RANGE_ALL 4


@interface HistoryModelOperation()

@property int TIME_RANGE;
@property int targetDays;

@end

@implementation HistoryModelOperation


-(id)init{
    self = [super init];
    self.TIME_RANGE = TIME_RANGE_LAST_TWO_WEEKS;
    self.dataLabels = [[NSMutableArray alloc]init];
    return self;
}

-(NSDate*)startTime{
    NSDate *date = nil;
    
    switch (self.TIME_RANGE) {
            case TIME_RANGE_LAST_WEEK:{
                self.targetDays = 7;
                date = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitDay
                                                                value:-7
                                                               toDate:[NSDate date]
                                                              options:0];
            }
            break;
            case TIME_RANGE_LAST_TWO_WEEKS:{
                self.targetDays = 14;
                date = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitDay
                                                                value:-14
                                                               toDate:[NSDate date]
                                                              options:0];
                
            }
            break;
            case TIME_RANGE_LAST_MONTH:{
                self.targetDays = 30;
                date = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitDay
                                                                value:-30
                                                               toDate:[NSDate date]
                                                              options:0];
                
            }
            break;
            case TIME_RANGE_ALL:{
                self.targetDays = -1;
                date = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitYear
                                                                value:-100
                                                               toDate:[NSDate date]
                                                              options:0];
                
            }
            break;
        default:
            break;
            
    }
    return date;
}



-(NSArray*)getRawData:(NSManagedObjectContext*)context{
    NSArray* rawDataList = [DBDetection getDetectionList:context];
    NSMutableArray* filterList = [[NSMutableArray alloc]init];
    for(DetectionData* eachData in rawDataList){
        NSDate* date1 = [eachData detectionTime];
        NSDate* date2 = [self startTime]; 
        if ([[eachData detectionTime] compare:[self startTime]] != NSOrderedAscending ) {
            [filterList addObject:eachData];
        }
    }
    return filterList;
}

-(NSArray*)fillDummyIfNoEnoughData:(int)desiredLength currentLength:(int)currentLength data:(NSArray*)data{
    if (currentLength >= desiredLength) {
        return data;
    }
    NSMutableArray* newData = [[NSMutableArray alloc]init];
    for(int i = 0 ; i < desiredLength; i++){
        [newData addObject:@-1];
    }
    
    for (int i=0; i<[data count]; i++) {
        int index = currentLength-1-i;
        NSNumber* temp = [data objectAtIndex: index];
        newData[desiredLength-1-i] = temp;
    }
    
    for(int i = 0 ; i < [newData count]; i ++){
        if ([[newData objectAtIndex:i] intValue] < 0 || ![newData objectAtIndex:i]) {
            newData[i] = @0;
        }else{
            break;
        }
    }
    return newData;
}

-(NSDate*) firstDateOf:(NSArray*)list desiredTag:(NSString*)desiredTag{
    for(int i =0 ; i < [list count]; i++){
        DetectionData* data = [list objectAtIndex:i];
        if ([[data pre_post] isEqualToString:desiredTag]) {
            return [data detectionTime];
        }
    }
    return nil;
}


-(int)getTotalDays:(NSArray*)list desiredTag:(NSString*)desiredTag{
    NSDate* date = [self firstDateOf:list desiredTag:desiredTag];
    if (date!=nil) {
        return [Util countDaysDifferent:date toDate:[[NSDate alloc]init]];
    }
    return 0;
}





/**
 *
 *  Get data by Date
 *
 *  targetBodyPart : what part you looking for.  if null, means gets all parts
 *
 * */

-(NSDictionary*)getDataByDate:(NSManagedObjectContext*)context targetBodyPart:(NSString*)targetBodyPart{

    NSDate* targetFirstDate = [Util convertStringToDate:[Util getDate:[[NSDate alloc]init] offset:-self.targetDays] withFormat:FORMAT_DATE];
    // create labels
    self.dataLabels = [[NSMutableArray alloc]init];
    for(int i=0; i < self.targetDays; i ++){
        NSString* date = [Util getDate:targetFirstDate offset:i];
        if(i%2==0){
            self.dataLabels[i] = [Util convertDateToString:[Util convertStringToDate:date withFormat:FORMAT_DATE] withFormat:@"MM/dd"];
        }else{
            self.dataLabels[i] = @"";
        }
    }

    
    NSArray* rawDataList = [self getRawData:context];
    if (rawDataList && [rawDataList count] > 0 ) {
        NSDate* firstDatePre = [self firstDateOf:rawDataList desiredTag:PRE_NURSING];
        NSDate* firstDatePost = [self firstDateOf:rawDataList desiredTag:POST_NURSING];
        int totalDaysForPre = [self getTotalDays:rawDataList desiredTag:PRE_NURSING];
        int totalDaysForPost = [self getTotalDays:rawDataList desiredTag:POST_NURSING];
        
        
        NSMutableArray* preNursingRecord = [[NSMutableArray alloc]init];
        NSMutableArray* postNursingRecord = [[NSMutableArray alloc]init];
        for(int i = 0 ; i < totalDaysForPre; i++){
            [preNursingRecord addObject:@0];
        }
        for(int i = 0 ; i < totalDaysForPost; i++){
            [postNursingRecord addObject:@0];
        }
        
        for(int i =0; i < totalDaysForPre; i ++){
            // create data
            int preNursingSum = 0 ;
            int numPre = 0 ;
            NSString* date = [Util getDate:firstDatePre offset:i];
            for (DetectionData* eachData in rawDataList) {
                NSDate* tempDate = [eachData detectionTime];
                NSString* dateOnlyStr = [Util convertDateToString:tempDate withFormat:FORMAT_DATE];
                if([dateOnlyStr isEqualToString:date]){
                    if([PRE_NURSING isEqualToString:eachData.pre_post]){
                        if(targetBodyPart == nil || [targetBodyPart isEqualToString:eachData.bodyPart]){
                            preNursingSum += eachData.value.intValue;
                            numPre ++;
                        }
                    }
                }
            }
            if (numPre!=0) {
                preNursingRecord[i] = @(preNursingSum*1.0/numPre);
            }else{
                preNursingRecord[i] = @(0);
            }
        }
        
        // if value == 0 , use previous record , therefore avoid sudden drop on graphic
        for (int i=1 ; i < totalDaysForPre; i++) {
            if([[preNursingRecord objectAtIndex:i] intValue] == 0 ){
                preNursingRecord[i] = [preNursingRecord objectAtIndex:i-1];
            }
        }
        
        for(int i =0; i < totalDaysForPost; i ++){
            // create data
            int postNursingSum = 0 ;
            int numPost = 0 ;
            NSString* date = [Util getDate:firstDatePost offset:i];
            for (DetectionData* eachData in rawDataList) {
                NSDate* tempDate = [eachData detectionTime];
                NSString* dateOnlyStr = [Util convertDateToString:tempDate withFormat:FORMAT_DATE];
                if([dateOnlyStr isEqualToString:date]){
                    if([POST_NURSING isEqualToString:eachData.pre_post]){
                        if(targetBodyPart == nil || [targetBodyPart isEqualToString:eachData.bodyPart]){
                            postNursingSum += eachData.value.intValue;
                            numPost ++;
                        }
                    }
                }
            }
            if (numPost!=0) {
                postNursingRecord[i] = @(postNursingSum*1.0/numPost);
            }else{
                postNursingRecord[i] = @(0);
            }
        }
        
        // if value == 0 , use previous record , therefore avoid sudden drop on graphic
        for (int i=1 ; i < totalDaysForPost; i++) {
            if([[postNursingRecord objectAtIndex:i] intValue] == 0 ){
                postNursingRecord[i] = [postNursingRecord objectAtIndex:i-1];
            }
        }
        
        NSArray* preArr = [self fillDummyIfNoEnoughData:self.targetDays currentLength:totalDaysForPre data:preNursingRecord];
        NSArray* postArr = [self fillDummyIfNoEnoughData:self.targetDays currentLength:totalDaysForPost data:postNursingRecord];
        
   
        NSMutableDictionary* result = [[NSMutableDictionary alloc]init];
        [result setObject:preArr forKey:PRE_NURSING];
        [result setObject:postArr forKey:POST_NURSING];
        
        return result; 
        
    }
    return [self createEmptyRecord];
}


-(NSDictionary*)getDataByBodyPart:(NSManagedObjectContext*)context{
    NSArray* rawDataList = [self getRawData:context];
    NSMutableDictionary* record = [[NSMutableDictionary alloc]init];
    [record setObject:@0 forKey:HAND];
    [record setObject:@0 forKey:FACE];
    [record setObject:@0 forKey:EYES];
    
    for(DetectionData* eachData in rawDataList){
        if ([[eachData bodyPart]isEqualToString:HAND]) {
            NSNumber* curretVal = [record objectForKey:HAND];
            [record setObject:@(curretVal.intValue +1) forKey:HAND];
        }
        if ([[eachData bodyPart]isEqualToString:FACE]) {
            NSNumber* curretVal = [record objectForKey:FACE];
            [record setObject:@(curretVal.intValue +1) forKey:FACE];
        }
        if ([[eachData bodyPart]isEqualToString:EYES]) {
            NSNumber* curretVal = [record objectForKey:EYES];
            [record setObject:@(curretVal.intValue +1) forKey:EYES];
        }
    }
    return record;
}


-(NSDictionary*)createEmptyRecord{
    NSMutableArray* preNursingRecord = [[NSMutableArray alloc]init];
    NSMutableArray* postNursingRecord = [[NSMutableArray alloc]init];
    for(int i = 0 ; i < self.targetDays; i++){
        [preNursingRecord addObject:@0];
    }
    for(int i = 0 ; i < self.targetDays; i++){
        [postNursingRecord addObject:@0];
    }
    NSMutableDictionary* result = [[NSMutableDictionary alloc]init];
    [result setObject:preNursingRecord forKey:PRE_NURSING];
    [result setObject:postNursingRecord forKey:POST_NURSING];
    return result;
}

@end

File: android_v2/app/src/main/java/com/mlz/shuiguang/model/HistoryModelOperation.java
package com.mlz.shuiguang.model;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.provider.Contacts;

import com.github.mikephil.charting.data.Entry;
import com.mlz.shuiguang.database.operation.TBDetection;
import com.mlz.shuiguang.transmission.bmobObject.Detection;
import com.mlz.shuiguang.util.Util;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;


/**
 * Created by dehualai on 12/28/16.
 */

public class HistoryModelOperation {


    public static final int TIME_RANGE_LAST_WEEK = 1;
    public static final int TIME_RANGE_LAST_TWO_WEEKS = 2;
    public static final int TIME_RANGE_LAST_MONTH = 3;
    public static final int TIME_RANGE_ALL = 4;

    private int TIME_RANGE = TIME_RANGE_LAST_TWO_WEEKS;

    private static String[] dateLabels;
    private int targetDays = -1;



    private Date startTime(){
        Calendar calendar = Calendar.getInstance();
        switch (TIME_RANGE){
            case TIME_RANGE_LAST_WEEK:
                calendar.add(Calendar.DATE, -7);
                targetDays = 7;
                break;
            case TIME_RANGE_LAST_TWO_WEEKS:
                calendar.add(Calendar.DATE, -14);
                targetDays = 14;
                break;
            case TIME_RANGE_LAST_MONTH:
                calendar.add(Calendar.DATE, -30);
                targetDays = 30;
                break;
            case TIME_RANGE_ALL:
                calendar.add(Calendar.YEAR, -100); // 100 years before
                targetDays = -1;
                break;
            default:
                break;
        }
        return calendar.getTime();
    }




    ///////////////////////////  2017-02-15  //////////////////////////////

    private ArrayList<DetectionData> getRawData(SQLiteDatabase db){
        TBDetection tbDetection = new TBDetection();
        ArrayList<DetectionData> rawDataList = tbDetection.getDetectionList(db);
        ArrayList<DetectionData> filterList = new ArrayList<>();
        // step 1 : filter data list
        for (DetectionData eachData : rawDataList){
            if (!eachData.getDetectionTime().before(startTime())){
                filterList.add(eachData);
            }
        }
        return filterList;
    }





    private Integer[] fillDummyIfNoEnoughData(int desiredLength, int currentLength, Integer[] data){
        if (currentLength>=desiredLength){
            return data;
        }
        Integer[] newData = new Integer[desiredLength];
        for(int i = 0; i < data.length; i++){
            newData[desiredLength-1-i] = data[currentLength-1-i];
        }
        for(int i = 0 ; i < newData.length; i++){
            if (newData[i] == null || newData[i] <0){
                newData[i] = 0;
            }else{
                break;
            }
        }
        return newData;
    }


    private Date firstDateOf(ArrayList<DetectionData> list, String desiredTag){
        for(int i = 0 ; i < list.size(); i++){
            DetectionData data = list.get(i);
            if(data.getPre_post().equals(desiredTag)){
                return data.getDetectionTime();
            }
        }
        return null;
    }

    private int getTotalDays(ArrayList<DetectionData> list, String desiredTag){
        Date date = firstDateOf(list, desiredTag);
        if (date!=null){
            return Util.countDaysDifferent(date, new Date());
        }
        return 0;
    }

    public static String[] getDateLabels(){
        return dateLabels;
    }

    /**
     *
     *  Get data by Date
     *
     *  targetBodyPart : what part you looking for.  if null, means gets all parts
     *
     * */
    public HashMap<String, Integer[]> getDataByDate(SQLiteDatabase db, String targetBodyPart){
        ArrayList<DetectionData> rawDataList = getRawData(db);
        if (rawDataList!=null && rawDataList.size()>0){
            Date firstDatePre = firstDateOf(rawDataList, DetectionData.PRE_NURSING);
            Date firstDatePost = firstDateOf(rawDataList, DetectionData.POST_NURSING);
            int totalDaysForPre = getTotalDays(rawDataList, DetectionData.PRE_NURSING);
            int totalDaysForPost = getTotalDays(rawDataList, DetectionData.POST_NURSING);
            Date targetFirstDate =  Util.convertStringToTime(Util.getDate(new Date(), -targetDays), Util.FORMAT_DATE);

            Integer[] preNursingRecord = new Integer[totalDaysForPre];
            Integer[] postNursingRecord = new Integer[totalDaysForPost];




            // create labels
            dateLabels = new String[targetDays];
            for(int i=0 ; i < targetDays; i++){
                String date = Util.getDate(targetFirstDate, i);
                // create labels for xAxis
                if (i%2==0) {
                    dateLabels[i] = Util.convertTimeToString(Util.convertStringToTime(date, Util.FORMAT_DATE), "MM/dd");
                }else{
                    dateLabels[i] = "";
                }
            }

            for(int i=0; i < totalDaysForPre; i++){
                // create data
                int preNursingSum = 0;
                int numPre = 0 ;
                String date = Util.getDate(firstDatePre, i);

                for (DetectionData eachData : rawDataList){
                    if (Util.convertTimeToString(eachData.getDetectionTime(), Util.FORMAT_DATE).equals(date)){
                        if (DetectionData.PRE_NURSING.equals(eachData.getPre_post())){
                            if(targetBodyPart==null || targetBodyPart.equals(eachData.getBodyPart())){
                                preNursingSum += eachData.getValue();
                                numPre++;
                            }
                        }
                    }
                }

                if (numPre!=0){
                    preNursingRecord[i] = (int)(preNursingSum*1.0f/numPre);
                }else{
                    preNursingRecord[i] = 0 ;
                }
            }

            // if value == 0 , use previous record , therefore avoid sudden drop on graphic
            for(int i=1 ; i < totalDaysForPre; i++){
                if(preNursingRecord[i]==0){
                    preNursingRecord[i] = preNursingRecord[i-1];
                }
            }


            for(int i=0; i < totalDaysForPost; i++){
                // create data
                int postNursingSum = 0;
                int numPost = 0 ;
                String date = Util.getDate(firstDatePost, i);

                for (DetectionData eachData : rawDataList){
                    if (Util.convertTimeToString(eachData.getDetectionTime(), Util.FORMAT_DATE).equals(date)){
                        if (DetectionData.POST_NURSING.equals(eachData.getPre_post())){
                            if(targetBodyPart==null || targetBodyPart.equals(eachData.getBodyPart())){
                                postNursingSum += eachData.getValue();
                                numPost++;
                            }
                        }
                    }
                }

                if (numPost!=0){
                    postNursingRecord[i] = (int)(postNursingSum*1.0f/numPost);
                }else{
                    postNursingRecord[i] = 0 ;
                }
            }


            // if value == 0 , use previous record , therefore avoid sudden drop on graphic
            for(int i=1 ; i < totalDaysForPost; i++){
                if(postNursingRecord[i]==0){
                    postNursingRecord[i] = postNursingRecord[i-1];
                }
            }


            // make list long
            preNursingRecord = fillDummyIfNoEnoughData(targetDays, totalDaysForPre, preNursingRecord);
            postNursingRecord = fillDummyIfNoEnoughData(targetDays, totalDaysForPost, postNursingRecord);



            HashMap<String, Integer[]> result = new HashMap<String, Integer[]>();
            result.put(DetectionData.PRE_NURSING, preNursingRecord);
            result.put(DetectionData.POST_NURSING, postNursingRecord);
            return result;
        }
        return null;
    }

    public HashMap<String, Integer> getDataByBodyPart(SQLiteDatabase db){

        ArrayList<DetectionData> rawDataList = getRawData(db);

        HashMap<String, Integer> record = new HashMap<>();
        record.put(DetectionData.HAND, 0);
        record.put(DetectionData.FACE, 0);
        record.put(DetectionData.EYES, 0);

        for (DetectionData eachData : rawDataList){
            if(DetectionData.HAND.equals(eachData.getBodyPart())){
                record.put(DetectionData.HAND, record.get(DetectionData.HAND)+1);
            }
            if(DetectionData.FACE.equals(eachData.getBodyPart())){
                record.put(DetectionData.FACE, record.get(DetectionData.FACE)+1);
            }
            if(DetectionData.EYES.equals(eachData.getBodyPart())){
                record.put(DetectionData.EYES, record.get(DetectionData.EYES)+1);
            }
        }

        return record;

    }






}
